### PR TITLE
Removed linear gradient heap allocation for color conversions between dart and display list

### PR DESCRIPTION
**Pull request by gaaclarke**
issue: https://github.com/flutter/flutter/issues/154650

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/engine/blob/main/docs/testing/Testing-the-engine.md
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md

### CODE REVIEW

**Review by jonahwilliams — COMMENTED** [review]
don't the other gradient types need these changes too?

File: display_list/effects/dl_color_source.cc [hunk 1]
```diff
@@ -157,23 +181,71 @@ bool DlGradientColorSourceBase::base_equals_(
...
       div = 1;
     }
-    for (uint32_t i = 0; i < stop_count_; i++) {
+    for (uint32_t i = 0; i < stop_count; i++) {
       stop_storage[i] = i / div;
     }
   }
 }
 
+class DlScalerToDlColorIt {
```
Comment (jonahwilliams) [inline display_list/effects/dl_color_source.cc:210]: Til how iterators work in C++.

Maybe `DlColorConversionIterator`?

File: display_list/effects/dl_color_source.cc [hunk 2]
```diff
@@ -38,9 +45,26 @@ std::shared_ptr<DlColorSource> DlColorSource::MakeLinear(
...
+                                            colors, stops, tile_mode, matrix),
+            DlGradientDeleter);
+  return ret;
+}
+
+std::shared_ptr<DlColorSource> DlColorSource::MakeLinear(
+    const DlPoint start_point,
+    const DlPoint end_point,
+    uint32_t stop_count,
+    const DlScalar* colors_argb,
```
Comment (flar) [inline display_list/effects/dl_color_source.cc:63]: Below this is referred to as just "colors"???

Comment (gaaclarke) [reply display_list/effects/dl_color_source.cc:63]: thanks, done

File: display_list/effects/dl_color_source.cc (hunk 1 above)
Comment (gaaclarke) [reply display_list/effects/dl_color_source.cc]: > Til how iterators work in C++.

Yea, just think of using operator overloading to make a class work like a pointer and use templates to accept anything that works like a pointer.

> Maybe DlColorConversionIterator?

done

**Comment by gaaclarke** [comment]
> don't the other gradient types need these changes too?

yep, they'll happen in different PRs

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Pull request merged**